File: MPI/exercises/ex8/vector_type.f90
program datatype1

  use mpi

  implicit none

  integer, dimension(8,8) :: array
  integer :: rank, ierr
  integer :: rowtype
  integer :: i, j

  call mpi_init(ierr)
  call mpi_comm_rank(MPI_COMM_WORLD, rank ,ierr)

  ! initialize arrays
  if (rank == 0) then
     do i=1,8
        do j=1,8
           array(i,j) = i*10 + j
        end do
     end do
  else
     array(:,:) = 0
  end if

  ! TODO start
  ! create datatype


  ! send first row of matrix
  if (rank == 0) then
     call mpi_send( ,  , , 1, 1, MPI_COMM_WORLD, ierr)
  else if (rank == 1) then
     call mpi_recv( ,  , , 0, 1, MPI_COMM_WORLD, MPI_STATUS_IGNORE, &
          ierr)
  end if

  ! TODO end

  ! Print out the result
  if (rank == 1) then
     do i=1,8
        write(*,'(8I3)') array(i, :)
     end do
  end if

  ! TODO free datatype
  call mpi_finalize(ierr)

end program datatype1

File: MPI/exercises/ex8/vector_type.c
#include <stdio.h>
#include <mpi.h>

int main(int argc, char **argv)
{
    int rank;
    int array[8][8];
    MPI_Datatype columntype;

    int i, j;

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    // Initialize arrays
    if (rank == 0) {
        for (i = 0; i < 8; i++) {
            for (j = 0; j < 8; j++) {
                array[i][j] = (i + 1) * 10 + j + 1;
            }
        }
    } else {
        for (i = 0; i < 8; i++) {
            for (j = 0; j < 8; j++) {
                array[i][j] = 0;
            }
        }
    }

    /* TODO start */
    // Create datatype

    // Send first column of matrix
    if (rank == 0) {
        MPI_Send( , , , 1, 1, MPI_COMM_WORLD);
    } else if (rank == 1) {
        MPI_Recv( , , , 0, 1, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
    }
    /* TODO end */

    // Print out the result
    if (rank == 1) {
        for (i = 0; i < 8; i++) {
            for (j = 0; j < 8; j++) {
                printf("%3d", array[i][j]);
            }
            printf("\n");
        }
    }

    /* TODO: Free datatype*/
    MPI_Finalize();

    return 0;
}
